include/nbl/builtin/hlsl/workgroup2/arithmetic_config.hlsl

```diff
@@ -46,39 +50,72 @@ struct ArithmeticConfiguration
         0,
         conditional_value<LevelCount==3,uint16_t,
-            SubgroupSize*ItemsPerInvocation_2,
-            0
-            >::value + SubgroupSize*ItemsPerInvocation_1
+            SubgroupSize*ItemsPerInvocation_2+__ItemsPerVirtualWorkgroup,
+            SubgroupSize*ItemsPerInvocation_1
+            >::value
         >::value;
 
     static bool electLast()
     {
         return glsl::gl_SubgroupInvocationID()==SubgroupSize-1;
     }
 
-    static uint32_t virtualSubgroupID(const uint32_t subgroupID, const uint32_t virtualIdx)
+    // gets a subgroupID as if each workgroup has (VirtualWorkgroupSize/SubgroupSize) subgroups
+    // each subgroup does work (VirtualWorkgroupSize/WorkgroupSize) times, the index denoted by workgroupInVirtualIndex
+    static uint32_t virtualSubgroupID(const uint32_t subgroupID, const uint32_t workgroupInVirtualIndex)
+    {
+        return workgroupInVirtualIndex * (WorkgroupSize >> SubgroupSizeLog2) + subgroupID;
+    }
+
+    // get a coalesced index to store for the next level in shared mem, e.g. level 0 -> level 1
+    // specify the next level to store values for in template param
+    // at level==LevelCount-1, it is guaranteed to have SubgroupSize elements
+    template<uint16_t level>
+    static uint32_t sharedStoreIndex(const uint32_t subgroupID)
     {
-        return virtualIdx * (WorkgroupSize >> SubgroupSizeLog2) + subgroupID;
+        uint32_t offsetBySubgroup;
+        if (level == LevelCount-1)
+            offsetBySubgroup = SubgroupSize;
+        else
+            offsetBySubgroup = __SubgroupsPerVirtualWorkgroup;
+
+        if (level<2)
+            return (subgroupID & (ItemsPerInvocation_1-1)) * offsetBySubgroup + (subgroupID/ItemsPerInvocation_1);
+        else
+            return (subgroupID & (ItemsPerInvocation_2-1)) * offsetBySubgroup + (subgroupID/ItemsPerInvocation_2);
     }
 
-    static uint32_t sharedCoalescedIndexNextLevel(const uint32_t subgroupID, const uint32_t itemsPerInvocation)
+    template<uint16_t level>
+    static uint32_t sharedStoreIndexFromVirtualIndex(const uint32_t subgroupID, const uint32_t workgroupInVirtualIndex)
     {
-        return (subgroupID & (itemsPerInvocation-1)) * SubgroupSize + (subgroupID/itemsPerInvocation);
+        const uint32_t virtualID = virtualSubgroupID(subgroupID, workgroupInVirtualIndex);
+        return sharedStoreIndex<level>(virtualID);
     }
 
-    static uint32_t sharedCoalescedIndexByComponent(const uint32_t invocationIndex, const uint32_t component)
+    // get the coalesced index in shared mem at the current level
+    template<uint16_t level>
+    static uint32_t sharedLoadIndex(const uint32_t invocationIndex, const uint32_t component)
     {
-        return component * SubgroupSize + invocationIndex;
+        if (level == LevelCount-1)
+            return component * SubgroupSize + invocationIndex;
+        else
+            return component * __SubgroupsPerVirtualWorkgroup + invocationIndex;
```

Comment:
write it as `return component*(virtualSubgroupCount+1)+invocationIndex+invocationIndex/(SubgroupSize*ttemsPerInvocation);`

where `virtualSubgroupCount` and `ttemsPerInvocation` are the values for `level` obviously